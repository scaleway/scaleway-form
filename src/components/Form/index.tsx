import type { Decorator } from 'final-form'
import arrayMutators from 'final-form-arrays'
import createDecorator from 'final-form-focus'
import type { ReactNode } from 'react'
import type {
  FormRenderProps,
  FormProps as ReactFinalFormProps,
} from 'react-final-form'
import { Form as ReactFinalForm } from 'react-final-form'
import { ErrorProvider } from '../../providers'
import type { FormErrors } from '../../types'

const focusOnErrors = createDecorator()

export type FormProps<FormValues = unknown> = {
  children?:
    | ((props: FormRenderProps<FormValues, Partial<FormValues>>) => ReactNode)
    | ReactNode
  errors: FormErrors
  /**
   * onRawSubmit is the base onSubmit from final-form
   */
  onRawSubmit: ReactFinalFormProps<FormValues, Partial<FormValues>>['onSubmit']
  initialValues?: Partial<FormValues>
  validateOnBlur?: ReactFinalFormProps<
    FormValues,
    Partial<FormValues>
  >['validateOnBlur']
  validate?: ReactFinalFormProps<FormValues, Partial<FormValues>>['validate']
  /**
   * The form name attribute
   */
  name?: string
  render?: ReactFinalFormProps<FormValues, Partial<FormValues>>['render']
  mutators?: ReactFinalFormProps<FormValues, Partial<FormValues>>['mutators']
  keepDirtyOnReinitialize?: boolean
}

export const Form = <FormValues,>({
  children,
  onRawSubmit,
  errors,
  initialValues,
  validateOnBlur,
  validate,
  name,
  render,
  mutators,
  keepDirtyOnReinitialize,
}: FormProps<FormValues>): JSX.Element => (
  <ReactFinalForm
    initialValues={initialValues}
    validateOnBlur={validateOnBlur}
    validate={validate}
    decorators={[
      focusOnErrors as unknown as Decorator<FormValues, Partial<FormValues>>,
    ]}
    mutators={{
      ...arrayMutators,
      ...mutators,
    }}
    onSubmit={onRawSubmit}
    render={
      render ??
      (renderProps => (
        <ErrorProvider errors={errors}>
          <form noValidate name={name} onSubmit={renderProps.handleSubmit}>
            {typeof children === 'function' ? children(renderProps) : children}
          </form>
        </ErrorProvider>
      ))
    }
    keepDirtyOnReinitialize={keepDirtyOnReinitialize}
  />
)
