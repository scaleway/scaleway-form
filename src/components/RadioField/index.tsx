import { Radio } from '@scaleway/ui'
import type { FieldState } from 'final-form'
import type { ComponentProps, ReactNode } from 'react'
import { useFormField } from '../../hooks'
import { useErrors } from '../../providers/ErrorContext'
import type { BaseFieldProps } from '../../types'

type RadioValue = NonNullable<ComponentProps<typeof Radio>['value']>

type RadioFieldProps<T = RadioValue, K = string> = BaseFieldProps<T, K> &
  Partial<
    Pick<
      ComponentProps<typeof Radio>,
      'disabled' | 'id' | 'onBlur' | 'onChange' | 'onFocus' | 'size' | 'value'
    >
  > & {
    children?: ReactNode
    className?: string
    label?: string
    name: string
    required?: boolean
  }

export const RadioField = ({
  children,
  className,
  disabled,
  id,
  label = '',
  name,
  onBlur,
  onChange,
  onFocus,
  required,
  size,
  validate,
  value,
}: RadioFieldProps): JSX.Element => {
  const { getError } = useErrors()

  const { input, meta } = useFormField(name, {
    required,
    type: 'radio',
    validate,
    value,
  })

  const error = getError({
    disabled,
    label,
    meta: meta as FieldState<unknown>,
    name,
    value: input.value,
  })

  return (
    <Radio
      checked={input.checked}
      className={className}
      disabled={disabled}
      error={error}
      id={id}
      name={input.name}
      onChange={event => {
        input.onChange(event)
        onChange?.(event)
      }}
      onBlur={event => {
        input.onBlur(event)
        onBlur?.(event)
      }}
      onFocus={event => {
        input.onFocus(event)
        onFocus?.(event)
      }}
      required={required}
      size={size}
      value={input.value}
    >
      {children}
    </Radio>
  )
}
