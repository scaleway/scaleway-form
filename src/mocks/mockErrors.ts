import type { FormErrors } from '../types'

export const emailRegex = /^[A-Z0-9._%+-]+@[A-Z0-9.-]+\.[A-Z]{2,}$/i

export const mockErrors: FormErrors = {
  MAX_DATE: ({ maxDate }) => `Date must be lower than ${maxDate?.toString()}`,
  MAX_LENGTH: ({ maxLength }) =>
    `This field should have a length lower than ${maxLength}`,
  MIN_DATE: ({ minDate }) => `Date must be greater than ${minDate?.toString()}`,
  MIN_LENGTH: ({ minLength }) =>
    `This field should have a length greater than ${minLength}`,
  REGEX: ({ regex }) => {
    if (
      (Array.isArray(regex[0])
        ? regex?.[0]?.[0]?.source
        : regex?.[0]?.source) === emailRegex.source
    ) {
      return 'Invalid email'
    }

    return `This field should match the regex ${regex
      .map(r =>
        Array.isArray(r)
          ? r.map(nestedRegex => nestedRegex.source).join(' or ')
          : r.source,
      )
      .join(' and ')}`
  },
  REQUIRED: 'This field is required',
  TOO_HIGH: ({ max }) => `This field is too high (maximum is : ${max})`,
  TOO_LOW: 'This field is too low',
}
