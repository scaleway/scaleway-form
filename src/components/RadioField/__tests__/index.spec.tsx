import userEvent from '@testing-library/user-event'
import React from 'react'
import RadioField from '..'
import {
  shouldMatchEmotionSnapshot,
  shouldMatchEmotionSnapshotFormWrapper,
} from '../../../helpers/jestHelpers'
import mockErrors from '../../../mocks/mockErrors'
import Form from '../../Form'

describe('RadioField', () => {
  test('should render correctly', () =>
    shouldMatchEmotionSnapshotFormWrapper(
      <RadioField name="test" value="test">
        Radio field
      </RadioField>,
    ))

  test('should render correctly disabled', () =>
    shouldMatchEmotionSnapshotFormWrapper(
      <RadioField name="test" value="disabled" disabled>
        Radio field disabled
      </RadioField>,
      {
        transform: node => {
          const input = node.getByRole('radio')
          expect(input).toBeDisabled()
        },
      },
    ))

  test('should render correctly checked', () =>
    shouldMatchEmotionSnapshot(
      <Form errors={mockErrors} initialValues={{ test: 'checked' }}>
        <RadioField name="test" value="checked">
          Radio field checked
        </RadioField>
      </Form>,
      {
        transform: node => {
          const input = node.getByRole('radio')
          expect(input).toBeChecked()
        },
      },
    ))

  test('should trigger events correctly', () => {
    const onFocus = jest.fn(() => {})
    const onChange = jest.fn(() => {})
    const onBlur = jest.fn(() => {})

    return shouldMatchEmotionSnapshotFormWrapper(
      <RadioField
        name="test"
        value="events"
        onChange={onChange}
        onFocus={onFocus}
        onBlur={onBlur}
      >
        Radio field events
      </RadioField>,
      {
        transform: node => {
          const input = node.getByRole('radio')
          input.focus()
          expect(onFocus).toBeCalledTimes(1)
          input.click()
          expect(onChange).toBeCalledTimes(1)
          input.blur()
          expect(onBlur).toBeCalledTimes(1)
        },
      },
    )
  })

  test('should render correctly with errors', () =>
    shouldMatchEmotionSnapshot(
      <Form errors={mockErrors}>
        <RadioField name="test" value="checked" required>
          Radio field error
        </RadioField>
        <button type="submit">Submit</button>
      </Form>,
      {
        transform: async node => {
          await userEvent.click(node.getByRole('button'))
          const error = node.getByText(mockErrors.REQUIRED as string)
          expect(error).toBeVisible()
        },
      },
    ))
})
