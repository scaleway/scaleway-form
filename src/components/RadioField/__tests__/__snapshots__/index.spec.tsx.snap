// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RadioField should render correctly 1`] = `
<DocumentFragment>
  .cache-1fdj0xh {
  position: relative;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: flex-start;
  -webkit-box-align: flex-start;
  -ms-flex-align: flex-start;
  align-items: flex-start;
  gap: 8px;
}

.cache-1fdj0xh[aria-disabled='false'] {
  cursor: pointer;
}

.cache-1fdj0xh:hover[aria-disabled='false'] .e1duuirp1[id=test-test]+.e1duuirp2 {
  background-color: #eeeeff;
  fill: #4f0599;
}

.cache-1fdj0xh:hover[aria-disabled='false'] .e1duuirp1[id=test-test]+.e1duuirp2 .e1duuirp4 {
  fill: #eeeeff;
}

.cache-1fdj0xh:hover[aria-disabled='false'] .e1duuirp1[id=test-test][aria-invalid='true']+.e1duuirp2 {
  background-color: #ffe1e7;
  fill: #a6102d;
}

.cache-1fdj0xh:hover[aria-disabled='false'] .e1duuirp1[id=test-test][aria-invalid='true']+.e1duuirp2 .e1duuirp4 {
  fill: #ffe1e7;
}

.cache-1fdj0xh[aria-disabled='true'] {
  cursor: not-allowed;
  color: #c4c9d6;
}

.cache-1fdj0xh[aria-disabled='true'] .e1duuirp2 {
  fill: #c4c9d6;
}

.cache-1fdj0xh[aria-disabled='true'] .e1duuirp2 .e1duuirp4 {
  fill: #f6f6f8;
}

.cache-1n4xajr {
  cursor: pointer;
  position: absolute;
  height: 24px;
  width: 24px;
  opacity: 0;
  white-space: nowrap;
  border-width: 0;
}

.cache-1n4xajr+.e1duuirp2 .e1duuirp3 {
  transform-origin: center;
  -webkit-transition: 200ms -webkit-transform ease-in-out;
  transition: 200ms transform ease-in-out;
  -webkit-transform: scale(0);
  -moz-transform: scale(0);
  -ms-transform: scale(0);
  transform: scale(0);
}

.cache-1n4xajr:checked+svg .e1duuirp3 {
  -webkit-transform: scale(1);
  -moz-transform: scale(1);
  -ms-transform: scale(1);
  transform: scale(1);
}

.cache-1n4xajr:checked[aria-disabled='false'][aria-invalid='false']+.e1duuirp2 {
  fill: #4f0599;
}

.cache-1n4xajr[aria-invalid='true']+.e1duuirp2 {
  fill: #a6102d;
}

.cache-1n4xajr:focus+.e1duuirp2 {
  background-color: #eeeeff;
  fill: #4f0599;
}

.cache-1n4xajr:focus+.e1duuirp2 .e1duuirp4 {
  fill: #eeeeff;
}

.cache-1n4xajr[aria-invalid='true']:focus+.e1duuirp2 {
  background-color: #ffe1e7;
  fill: #a6102d;
}

.cache-1n4xajr[aria-invalid='true']:focus+.e1duuirp2 .e1duuirp4 {
  fill: #ffe1e7;
}

.cache-jsk10b {
  height: 24px;
  width: 24px;
  min-width: 24px;
  min-height: 24px;
  border-radius: 100%;
  fill: #b2b6c3;
}

.cache-jsk10b .e1duuirp4 {
  fill: #ffffff;
}

.cache-1e98oqk {
  fill: #ffffff;
}

<form
    novalidate=""
  >
    <label
      aria-disabled="false"
      class="cache-1fdj0xh e1duuirp0"
      data-checked="false"
      for="test-test"
    >
      <input
        aria-disabled="false"
        aria-invalid="false"
        class="cache-1n4xajr e1duuirp1"
        id="test-test"
        name="test"
        type="radio"
        value="test"
      />
      <svg
        class="cache-jsk10b e1duuirp2"
        size="24"
        viewBox="0 0 24 24"
      >
        <g>
          <circle
            cx="12"
            cy="12"
            r="8"
            stroke-width="2"
          />
          <circle
            class="cache-1e98oqk e1duuirp4"
            cx="12"
            cy="12"
            r="6"
          />
          <circle
            class="cache-0 e1duuirp3"
            cx="12"
            cy="12"
            r="4"
          />
        </g>
      </svg>
      Radio field
    </label>
  </form>
</DocumentFragment>
`;

exports[`RadioField should render correctly checked 1`] = `
<DocumentFragment>
  .cache-1imu6pm {
  position: relative;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: flex-start;
  -webkit-box-align: flex-start;
  -ms-flex-align: flex-start;
  align-items: flex-start;
  gap: 8px;
}

.cache-1imu6pm[aria-disabled='false'] {
  cursor: pointer;
}

.cache-1imu6pm:hover[aria-disabled='false'] .e1duuirp1[id=test-checked]+.e1duuirp2 {
  background-color: #eeeeff;
  fill: #4f0599;
}

.cache-1imu6pm:hover[aria-disabled='false'] .e1duuirp1[id=test-checked]+.e1duuirp2 .e1duuirp4 {
  fill: #eeeeff;
}

.cache-1imu6pm:hover[aria-disabled='false'] .e1duuirp1[id=test-checked][aria-invalid='true']+.e1duuirp2 {
  background-color: #ffe1e7;
  fill: #a6102d;
}

.cache-1imu6pm:hover[aria-disabled='false'] .e1duuirp1[id=test-checked][aria-invalid='true']+.e1duuirp2 .e1duuirp4 {
  fill: #ffe1e7;
}

.cache-1imu6pm[aria-disabled='true'] {
  cursor: not-allowed;
  color: #c4c9d6;
}

.cache-1imu6pm[aria-disabled='true'] .e1duuirp2 {
  fill: #c4c9d6;
}

.cache-1imu6pm[aria-disabled='true'] .e1duuirp2 .e1duuirp4 {
  fill: #f6f6f8;
}

.cache-1n4xajr {
  cursor: pointer;
  position: absolute;
  height: 24px;
  width: 24px;
  opacity: 0;
  white-space: nowrap;
  border-width: 0;
}

.cache-1n4xajr+.e1duuirp2 .e1duuirp3 {
  transform-origin: center;
  -webkit-transition: 200ms -webkit-transform ease-in-out;
  transition: 200ms transform ease-in-out;
  -webkit-transform: scale(0);
  -moz-transform: scale(0);
  -ms-transform: scale(0);
  transform: scale(0);
}

.cache-1n4xajr:checked+svg .e1duuirp3 {
  -webkit-transform: scale(1);
  -moz-transform: scale(1);
  -ms-transform: scale(1);
  transform: scale(1);
}

.cache-1n4xajr:checked[aria-disabled='false'][aria-invalid='false']+.e1duuirp2 {
  fill: #4f0599;
}

.cache-1n4xajr[aria-invalid='true']+.e1duuirp2 {
  fill: #a6102d;
}

.cache-1n4xajr:focus+.e1duuirp2 {
  background-color: #eeeeff;
  fill: #4f0599;
}

.cache-1n4xajr:focus+.e1duuirp2 .e1duuirp4 {
  fill: #eeeeff;
}

.cache-1n4xajr[aria-invalid='true']:focus+.e1duuirp2 {
  background-color: #ffe1e7;
  fill: #a6102d;
}

.cache-1n4xajr[aria-invalid='true']:focus+.e1duuirp2 .e1duuirp4 {
  fill: #ffe1e7;
}

.cache-jsk10b {
  height: 24px;
  width: 24px;
  min-width: 24px;
  min-height: 24px;
  border-radius: 100%;
  fill: #b2b6c3;
}

.cache-jsk10b .e1duuirp4 {
  fill: #ffffff;
}

.cache-1e98oqk {
  fill: #ffffff;
}

<form
    novalidate=""
  >
    <label
      aria-disabled="false"
      class="cache-1imu6pm e1duuirp0"
      data-checked="true"
      for="test-checked"
    >
      <input
        aria-disabled="false"
        aria-invalid="false"
        checked=""
        class="cache-1n4xajr e1duuirp1"
        id="test-checked"
        name="test"
        type="radio"
        value="checked"
      />
      <svg
        class="cache-jsk10b e1duuirp2"
        size="24"
        viewBox="0 0 24 24"
      >
        <g>
          <circle
            cx="12"
            cy="12"
            r="8"
            stroke-width="2"
          />
          <circle
            class="cache-1e98oqk e1duuirp4"
            cx="12"
            cy="12"
            r="6"
          />
          <circle
            class="cache-0 e1duuirp3"
            cx="12"
            cy="12"
            r="4"
          />
        </g>
      </svg>
      Radio field checked
    </label>
  </form>
</DocumentFragment>
`;

exports[`RadioField should render correctly disabled 1`] = `
<DocumentFragment>
  .cache-1o5kav0 {
  position: relative;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: flex-start;
  -webkit-box-align: flex-start;
  -ms-flex-align: flex-start;
  align-items: flex-start;
  gap: 8px;
}

.cache-1o5kav0[aria-disabled='false'] {
  cursor: pointer;
}

.cache-1o5kav0:hover[aria-disabled='false'] .e1duuirp1[id=test-disabled]+.e1duuirp2 {
  background-color: #eeeeff;
  fill: #4f0599;
}

.cache-1o5kav0:hover[aria-disabled='false'] .e1duuirp1[id=test-disabled]+.e1duuirp2 .e1duuirp4 {
  fill: #eeeeff;
}

.cache-1o5kav0:hover[aria-disabled='false'] .e1duuirp1[id=test-disabled][aria-invalid='true']+.e1duuirp2 {
  background-color: #ffe1e7;
  fill: #a6102d;
}

.cache-1o5kav0:hover[aria-disabled='false'] .e1duuirp1[id=test-disabled][aria-invalid='true']+.e1duuirp2 .e1duuirp4 {
  fill: #ffe1e7;
}

.cache-1o5kav0[aria-disabled='true'] {
  cursor: not-allowed;
  color: #c4c9d6;
}

.cache-1o5kav0[aria-disabled='true'] .e1duuirp2 {
  fill: #c4c9d6;
}

.cache-1o5kav0[aria-disabled='true'] .e1duuirp2 .e1duuirp4 {
  fill: #f6f6f8;
}

.cache-1n4xajr {
  cursor: pointer;
  position: absolute;
  height: 24px;
  width: 24px;
  opacity: 0;
  white-space: nowrap;
  border-width: 0;
}

.cache-1n4xajr+.e1duuirp2 .e1duuirp3 {
  transform-origin: center;
  -webkit-transition: 200ms -webkit-transform ease-in-out;
  transition: 200ms transform ease-in-out;
  -webkit-transform: scale(0);
  -moz-transform: scale(0);
  -ms-transform: scale(0);
  transform: scale(0);
}

.cache-1n4xajr:checked+svg .e1duuirp3 {
  -webkit-transform: scale(1);
  -moz-transform: scale(1);
  -ms-transform: scale(1);
  transform: scale(1);
}

.cache-1n4xajr:checked[aria-disabled='false'][aria-invalid='false']+.e1duuirp2 {
  fill: #4f0599;
}

.cache-1n4xajr[aria-invalid='true']+.e1duuirp2 {
  fill: #a6102d;
}

.cache-1n4xajr:focus+.e1duuirp2 {
  background-color: #eeeeff;
  fill: #4f0599;
}

.cache-1n4xajr:focus+.e1duuirp2 .e1duuirp4 {
  fill: #eeeeff;
}

.cache-1n4xajr[aria-invalid='true']:focus+.e1duuirp2 {
  background-color: #ffe1e7;
  fill: #a6102d;
}

.cache-1n4xajr[aria-invalid='true']:focus+.e1duuirp2 .e1duuirp4 {
  fill: #ffe1e7;
}

.cache-jsk10b {
  height: 24px;
  width: 24px;
  min-width: 24px;
  min-height: 24px;
  border-radius: 100%;
  fill: #b2b6c3;
}

.cache-jsk10b .e1duuirp4 {
  fill: #ffffff;
}

.cache-1e98oqk {
  fill: #ffffff;
}

<form
    novalidate=""
  >
    <label
      aria-disabled="true"
      class="cache-1o5kav0 e1duuirp0"
      data-checked="false"
      for="test-disabled"
    >
      <input
        aria-disabled="true"
        aria-invalid="false"
        class="cache-1n4xajr e1duuirp1"
        disabled=""
        id="test-disabled"
        name="test"
        type="radio"
        value="disabled"
      />
      <svg
        class="cache-jsk10b e1duuirp2"
        size="24"
        viewBox="0 0 24 24"
      >
        <g>
          <circle
            cx="12"
            cy="12"
            r="8"
            stroke-width="2"
          />
          <circle
            class="cache-1e98oqk e1duuirp4"
            cx="12"
            cy="12"
            r="6"
          />
          <circle
            class="cache-0 e1duuirp3"
            cx="12"
            cy="12"
            r="4"
          />
        </g>
      </svg>
      Radio field disabled
    </label>
  </form>
</DocumentFragment>
`;

exports[`RadioField should trigger events correctly 1`] = `
<DocumentFragment>
  .cache-1o6cmnv {
  position: relative;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: flex-start;
  -webkit-box-align: flex-start;
  -ms-flex-align: flex-start;
  align-items: flex-start;
  gap: 8px;
}

.cache-1o6cmnv[aria-disabled='false'] {
  cursor: pointer;
}

.cache-1o6cmnv:hover[aria-disabled='false'] .e1duuirp1[id=test-events]+.e1duuirp2 {
  background-color: #eeeeff;
  fill: #4f0599;
}

.cache-1o6cmnv:hover[aria-disabled='false'] .e1duuirp1[id=test-events]+.e1duuirp2 .e1duuirp4 {
  fill: #eeeeff;
}

.cache-1o6cmnv:hover[aria-disabled='false'] .e1duuirp1[id=test-events][aria-invalid='true']+.e1duuirp2 {
  background-color: #ffe1e7;
  fill: #a6102d;
}

.cache-1o6cmnv:hover[aria-disabled='false'] .e1duuirp1[id=test-events][aria-invalid='true']+.e1duuirp2 .e1duuirp4 {
  fill: #ffe1e7;
}

.cache-1o6cmnv[aria-disabled='true'] {
  cursor: not-allowed;
  color: #c4c9d6;
}

.cache-1o6cmnv[aria-disabled='true'] .e1duuirp2 {
  fill: #c4c9d6;
}

.cache-1o6cmnv[aria-disabled='true'] .e1duuirp2 .e1duuirp4 {
  fill: #f6f6f8;
}

.cache-1n4xajr {
  cursor: pointer;
  position: absolute;
  height: 24px;
  width: 24px;
  opacity: 0;
  white-space: nowrap;
  border-width: 0;
}

.cache-1n4xajr+.e1duuirp2 .e1duuirp3 {
  transform-origin: center;
  -webkit-transition: 200ms -webkit-transform ease-in-out;
  transition: 200ms transform ease-in-out;
  -webkit-transform: scale(0);
  -moz-transform: scale(0);
  -ms-transform: scale(0);
  transform: scale(0);
}

.cache-1n4xajr:checked+svg .e1duuirp3 {
  -webkit-transform: scale(1);
  -moz-transform: scale(1);
  -ms-transform: scale(1);
  transform: scale(1);
}

.cache-1n4xajr:checked[aria-disabled='false'][aria-invalid='false']+.e1duuirp2 {
  fill: #4f0599;
}

.cache-1n4xajr[aria-invalid='true']+.e1duuirp2 {
  fill: #a6102d;
}

.cache-1n4xajr:focus+.e1duuirp2 {
  background-color: #eeeeff;
  fill: #4f0599;
}

.cache-1n4xajr:focus+.e1duuirp2 .e1duuirp4 {
  fill: #eeeeff;
}

.cache-1n4xajr[aria-invalid='true']:focus+.e1duuirp2 {
  background-color: #ffe1e7;
  fill: #a6102d;
}

.cache-1n4xajr[aria-invalid='true']:focus+.e1duuirp2 .e1duuirp4 {
  fill: #ffe1e7;
}

.cache-jsk10b {
  height: 24px;
  width: 24px;
  min-width: 24px;
  min-height: 24px;
  border-radius: 100%;
  fill: #b2b6c3;
}

.cache-jsk10b .e1duuirp4 {
  fill: #ffffff;
}

.cache-1e98oqk {
  fill: #ffffff;
}

<form
    novalidate=""
  >
    <label
      aria-disabled="false"
      class="cache-1o6cmnv e1duuirp0"
      data-checked="true"
      for="test-events"
    >
      <input
        aria-disabled="false"
        aria-invalid="false"
        class="cache-1n4xajr e1duuirp1"
        id="test-events"
        name="test"
        type="radio"
        value="events"
      />
      <svg
        class="cache-jsk10b e1duuirp2"
        size="24"
        viewBox="0 0 24 24"
      >
        <g>
          <circle
            cx="12"
            cy="12"
            r="8"
            stroke-width="2"
          />
          <circle
            class="cache-1e98oqk e1duuirp4"
            cx="12"
            cy="12"
            r="6"
          />
          <circle
            class="cache-0 e1duuirp3"
            cx="12"
            cy="12"
            r="4"
          />
        </g>
      </svg>
      Radio field events
    </label>
  </form>
</DocumentFragment>
`;
